README.md
# rmc-pull-to-refresh
---

React Mobile PullToRefresh Component.


[![NPM version][npm-image]][npm-url]
[![build status][travis-image]][travis-url]
[![Test coverage][coveralls-image]][coveralls-url]
[![gemnasium deps][gemnasium-image]][gemnasium-url]
[![npm download][download-image]][download-url]

[npm-image]: http://img.shields.io/npm/v/rmc-pull-to-refresh.svg?style=flat-square
[npm-url]: http://npmjs.org/package/rmc-pull-to-refresh
[travis-image]: https://img.shields.io/travis/react-component/m-pull-to-refresh.svg?style=flat-square
[travis-url]: https://travis-ci.org/react-component/m-pull-to-refresh
[coveralls-image]: https://img.shields.io/coveralls/react-component/m-pull-to-refresh.svg?style=flat-square
[coveralls-url]: https://coveralls.io/r/react-component/m-pull-to-refresh?branch=master
[gemnasium-image]: http://img.shields.io/gemnasium/react-component/m-pull-to-refresh.svg?style=flat-square
[gemnasium-url]: https://gemnasium.com/react-component/m-pull-to-refresh
[node-image]: https://img.shields.io/badge/node.js-%3E=_0.10-green.svg?style=flat-square
[node-url]: http://nodejs.org/download/
[download-image]: https://img.shields.io/npm/dm/rmc-pull-to-refresh.svg?style=flat-square
[download-url]: https://npmjs.org/package/rmc-pull-to-refresh


## Screenshots

<img src="https://os.alipayobjects.com/rmsportal/rcLuiqtnDULPyzG.gif" width="288"/>


## Development

```
npm install
npm start
```

## Example

http://localhost:8899/examples/

online example: http://react-component.github.io/m-pull-to-refresh/


## install

[![rmc-pull-to-refresh](https://nodei.co/npm/rmc-pull-to-refresh.png)](https://npmjs.org/package/rmc-pull-to-refresh)


## Usage

see example

## API

### props

| name     | description    | type     | default      |
|----------|----------------|----------|--------------|
| direction  | pull direction, can be `up` or `down` | String | `down` |
| distanceToRefresh | distance to pull to refresh | number | 50  |
| refreshing | Whether the view should be indicating an active refresh | bool | false |
| onRefresh  | Called when the view starts refreshing. | () => void | - |
| indicator  | indicator config | Object | `{ activate: 'release', deactivate: 'pull', release: 'loading', finish: 'finish' }` |
| className | additional css class of root dom node | String | - |
| prefixCls | prefix class | String | 'rmc-pull-to-refresh' |
| damping | pull damping, suggest less than 200 | number | 100 |
| scale | damping scale | number | 0.6 |

## Test Case

```
npm test
```

## Coverage

```
npm run coverage
```

## License

rmc-pull-to-refresh is released under the MIT license.


HISTORY.md
# History

---

## 1.0.13

- prevent pull if x changed greater than 20rpx

## 1.0.12

- add props.scale for damping

## 1.0.11

- remove the unknown-prop warning
- add sensitivity for pull action

## 1.0.9

- simplify damping prop

## 1.0.8

- add custom damping prop

## 1.0.0 / 2017-09-30

**Completely rewritten**

- Remove zscroller dependency.
- Support web's native roll.
- Add `direction` `onRefresh` `indicator` props and Remove `icon` `loading` `loadingFunction` `scrollerOptions` props.


tests/index.js
// do not add tests to this file, add tests to other .spec.js
const req = require.context('.', false, /\.spec\.js$/);
req.keys().forEach(req);


tests/usage.spec.js
// add spec here!
import '../assets/index.less';
import expect from 'expect.js';
// import TestUtils from 'react-addons-test-utils';
import * as React from 'react';
import * as ReactDOM from 'react-dom';
import PullToRefresh from '../index';

describe('m-pull-to-refresh', () => {
  let instance;
  let div;
  beforeEach(() => {
    div = document.createElement('div');
    document.body.appendChild(div);
  });
  afterEach(() => {
    ReactDOM.unmountComponentAtNode(div);
    document.body.removeChild(div);
  });

  function loadingFunction() {
    return new Promise((resolve) => {
      setTimeout(() => {
        resolve();
      }, 500);
    });
  }

  it('should add css class of root dom node', () => {
    instance = ReactDOM.render(
      <PullToRefresh
        className="forTest"
        loadingFunction={loadingFunction}
        icon={
          <div style={{ marginTop: 15 }}>
            <div className="pull">
              pull down to refresh
            </div>
            <div className="release">
              release to refresh
            </div>
          </div>
        }
        loading={<div style={{ marginTop: 15 }}>Loading</div>}
        distanceToRefresh={40}
        contentClassName="for-test"
        style={{
          height: 300,
          textAlign: 'center',
          width: '90%',
          margin: '0 auto',
          border: '1px solid red',
        }}
      >
        <div style={{ height: 50 }}>Item 1</div>
      </PullToRefresh>, div
    );
    expect(ReactDOM.findDOMNode(instance).className.indexOf('forTest') !== -1).to.be(true);
  });
});


tests/PullToRefresh.spec.js
import React from 'react';
import { render } from 'enzyme';
import renderToJson from 'enzyme-to-json';
import PullToRefresh from '../src';

describe('basic', () => {
  it('base.', () => {
    const wrapper = render(
      <PullToRefresh>
        <div style={{ height: 50 }}>Item 1</div>
      </PullToRefresh>
    );
    expect(renderToJson(wrapper)).toMatchSnapshot();
  });
});


src/wptr.1.1.js
export default function WebPullToRefresh() {
  /**
   * Hold all of the default parameters for the module
   * @type {object}
   */
  const defaults = {
    // Number of pixels of panning until refresh
    distanceToRefresh: 70,

    // Pointer to function that does the loading and returns a promise
    loadingFunction: false,

    // Dragging resistance level
    resistance: 2.5,
  };

  /**
   * Hold all of the merged parameter and default module options
   * @type {object}
   */
  let options = {};

  /**
   * Pan event parameters
   * @type {object}
   */
  const pan = {
    enabled: false,
    distance: 0,
    startingPositionY: 0,
  };

  function bodyClassRemove() {
    options.containerEl.classList.remove(`${options.prefixCls}-reset`);
    options.containerEl.removeEventListener('transitionend', bodyClassRemove, false);
  }

  /**
   * Reset all elements to their starting positions before any paning took place.
   */
  function _doReset() {
    options.containerEl.classList.remove(`${options.prefixCls}-loading`);
    options.containerEl.classList.remove(`${options.prefixCls}-refresh`);
    options.containerEl.classList.add(`${options.prefixCls}-reset`);
    options.containerEl.addEventListener('transitionend', bodyClassRemove, false);
  }

  /**
   * Position content and refresh elements to show that loading is taking place.
   */
  function _doLoading() {
    options.containerEl.classList.add(`${options.prefixCls}-loading`);

    // If no valid loading function exists, just reset elements
    if (!options.loadingFunction) {
      return _doReset();
    }

    // The loading function should return a promise
    const loadingPromise = options.loadingFunction();

    // For UX continuity, make sure we show loading for at least one second before resetting
    setTimeout(() => {
      // Once actual loading is complete, reset pull to refresh
      loadingPromise.then(_doReset, _doReset);
    }, 1000);
  }

  /**
   * Initialize pull to refresh, hammer, and bind pan events.
   *
   * @param {object=} params - Setup parameters for pull to refresh
   */
  function init(params) {
    options = {
      ...params,
      distanceToRefresh: params.distanceToRefresh || defaults.distanceToRefresh,
      loadingFunction: params.loadingFunction || defaults.loadingFunction,
      resistance: params.resistance || defaults.resistance,
    };
  }

  /**
   * Set the CSS transform on the content element to move it on the screen.
   */
  function _setContentPan() {
    // Use transforms to smoothly animate elements on desktop and mobile devices
    options.contentEl.style.transform = options.contentEl.style.webkitTransform =
      `translate3d( 0, ${pan.distance}px, 0 )`;
    options.ptrEl.style.transform = options.ptrEl.style.webkitTransform =
      `translate3d( 0, ${pan.distance - options.ptrEl.offsetHeight}px, 0 )`;
  }

  /**
   * Set/remove the loading body class to show or hide the loading indicator after pull down.
   */
  function _setBodyClass() {
    if (pan.distance > options.distanceToRefresh) {
      options.containerEl.classList.add(`${options.prefixCls}-refresh`);
    } else {
      options.containerEl.classList.remove(`${options.prefixCls}-refresh`);
    }
  }

  /**
   * Determine whether pan events should apply based on scroll position on panstart
   */
  function onPanStart() {
    if (options.containerEl.classList.contains(`${options.prefixCls}-loading`)) {
      pan.enabled = false;
      return;
    }

    pan.startingPositionY = options.containerEl.scrollTop;

    if (pan.startingPositionY === 0) {
      pan.enabled = true;
    }
  }

  /**
   * Handle element on screen movement when the pandown events is firing.
   *
   * @param {object} e - Event object
   */
  function onPanDown(e) {
    if (!pan.enabled) {
      return;
    }

    e.preventDefault();
    pan.distance = e.distance / options.resistance;

    _setContentPan();
    _setBodyClass();
  }

  /**
   * Handle element on screen movement when the pandown events is firing.
   *
   * @param {object} e - Event object
   */
  function onPanUp(e) {
    if (!pan.enabled || pan.distance === 0) {
      return;
    }

    e.preventDefault();

    if (pan.distance < e.distance / options.resistance) {
      pan.distance = 0;
    } else {
      pan.distance = e.distance / options.resistance;
    }

    _setContentPan();
    _setBodyClass();
  }

  /**
   * Determine how to animate and position elements when the panend event fires.
   *
   * @param {object} e - Event object
   */
  function onPanEnd(e) {
    if (!pan.enabled || pan.distance === 0) {
      return;
    }

    e.preventDefault();

    options.contentEl.style.transform = options.contentEl.style.webkitTransform = '';
    options.ptrEl.style.transform = options.ptrEl.style.webkitTransform = '';

    if (options.containerEl.classList.contains(`${options.prefixCls}-refresh`)) {
      _doLoading();
    } else {
      _doReset();
    }

    pan.distance = 0;
    pan.enabled = false;
  }

  function onPan(e) {
    if (options.contentEl.scrollTop > 0) {
      return;
    }
    if (e.additionalEvent === 'pandown') {
      onPanDown(e);
    }
    if (e.additionalEvent === 'panup') {
      onPanUp(e);
    }
  }

  return {
    init,
    events: {
      onPanStart,
      onPan,
      onPanEnd,
    },
  };
}


examples/base1.js
webpackJsonp([0],{

/***/ 109:
/***/ (function(module, exports, __webpack_require__) {

module.exports = __webpack_require__(55);


/***/ }),

/***/ 55:
/***/ (function(module, __webpack_exports__, __webpack_require__) {

"use strict";
Object.defineProperty(__webpack_exports__, "__esModule", { value: true });
/* harmony import */ var __WEBPACK_IMPORTED_MODULE_0_babel_runtime_helpers_classCallCheck__ = __webpack_require__(20);
/* harmony import */ var __WEBPACK_IMPORTED_MODULE_0_babel_runtime_helpers_classCallCheck___default = __webpack_require__.n(__WEBPACK_IMPORTED_MODULE_0_babel_runtime_helpers_classCallCheck__);
/* harmony import */ var __WEBPACK_IMPORTED_MODULE_1_babel_runtime_helpers_createClass__ = __webpack_require__(21);
/* harmony import */ var __WEBPACK_IMPORTED_MODULE_1_babel_runtime_helpers_createClass___default = __webpack_require__.n(__WEBPACK_IMPORTED_MODULE_1_babel_runtime_helpers_createClass__);
/* harmony import */ var __WEBPACK_IMPORTED_MODULE_2_babel_runtime_helpers_possibleConstructorReturn__ = __webpack_require__(23);
/* harmony import */ var __WEBPACK_IMPORTED_MODULE_2_babel_runtime_helpers_possibleConstructorReturn___default = __webpack_require__.n(__WEBPACK_IMPORTED_MODULE_2_babel_runtime_helpers_possibleConstructorReturn__);
/* harmony import */ var __WEBPACK_IMPORTED_MODULE_3_babel_runtime_helpers_inherits__ = __webpack_require__(22);
/* harmony import */ var __WEBPACK_IMPORTED_MODULE_3_babel_runtime_helpers_inherits___default = __webpack_require__.n(__WEBPACK_IMPORTED_MODULE_3_babel_runtime_helpers_inherits__);
/* harmony import */ var __WEBPACK_IMPORTED_MODULE_4__assets_index_less__ = __webpack_require__(50);
/* harmony import */ var __WEBPACK_IMPORTED_MODULE_4__assets_index_less___default = __webpack_require__.n(__WEBPACK_IMPORTED_MODULE_4__assets_index_less__);
/* harmony import */ var __WEBPACK_IMPORTED_MODULE_5__src__ = __webpack_require__(37);
/* harmony import */ var __WEBPACK_IMPORTED_MODULE_6_react__ = __webpack_require__(13);
/* harmony import */ var __WEBPACK_IMPORTED_MODULE_6_react___default = __webpack_require__.n(__WEBPACK_IMPORTED_MODULE_6_react__);
/* harmony import */ var __WEBPACK_IMPORTED_MODULE_7_react_dom__ = __webpack_require__(52);
/* harmony import */ var __WEBPACK_IMPORTED_MODULE_7_react_dom___default = __webpack_require__.n(__WEBPACK_IMPORTED_MODULE_7_react_dom__);









var Lv = function (_React$Component) {
  __WEBPACK_IMPORTED_MODULE_3_babel_runtime_helpers_inherits___default()(Lv, _React$Component);

  function Lv() {
    __WEBPACK_IMPORTED_MODULE_0_babel_runtime_helpers_classCallCheck___default()(this, Lv);

    return __WEBPACK_IMPORTED_MODULE_2_babel_runtime_helpers_possibleConstructorReturn___default()(this, (Lv.__proto__ || Object.getPrototypeOf(Lv)).apply(this, arguments));
  }

  __WEBPACK_IMPORTED_MODULE_1_babel_runtime_helpers_createClass___default()(Lv, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      if (this.props.pullToRefresh) {
        this.forceUpdate();
      }
    }
  }, {
    key: 'render',
    value: function render() {
      var _this2 = this;

      var child = this.props.children;
      if (this.props.pullToRefresh) {
        child = __WEBPACK_IMPORTED_MODULE_6_react___default.a.cloneElement(this.props.pullToRefresh, {
          getScrollContainer: function getScrollContainer() {
            return _this2.lv;
          }
        }, child);
      }
      return __WEBPACK_IMPORTED_MODULE_6_react___default.a.createElement(
        'div',
        { ref: function ref(el) {
            return _this2.lv = el;
          }, style: { height: 200, border: 1, overflow: 'auto' } },
        child
      );
    }
  }]);

  return Lv;
}(__WEBPACK_IMPORTED_MODULE_6_react___default.a.Component);

var App = function (_React$Component2) {
  __WEBPACK_IMPORTED_MODULE_3_babel_runtime_helpers_inherits___default()(App, _React$Component2);

  function App() {
    var _ref;

    var _temp, _this3, _ret;

    __WEBPACK_IMPORTED_MODULE_0_babel_runtime_helpers_classCallCheck___default()(this, App);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this3 = __WEBPACK_IMPORTED_MODULE_2_babel_runtime_helpers_possibleConstructorReturn___default()(this, (_ref = App.__proto__ || Object.getPrototypeOf(App)).call.apply(_ref, [this].concat(args))), _this3), _this3.state = {
      refreshing: false
    }, _temp), __WEBPACK_IMPORTED_MODULE_2_babel_runtime_helpers_possibleConstructorReturn___default()(_this3, _ret);
  }

  __WEBPACK_IMPORTED_MODULE_1_babel_runtime_helpers_createClass___default()(App, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      document.body.style.overflowY = navigator.userAgent.match(/Android|iPhone|iPad|iPod/i) ? 'hidden' : 'auto';
    }
  }, {
    key: 'render',
    value: function render() {
      var _this4 = this;

      return __WEBPACK_IMPORTED_MODULE_6_react___default.a.createElement(
        Lv,
        {
          pullToRefresh: __WEBPACK_IMPORTED_MODULE_6_react___default.a.createElement(__WEBPACK_IMPORTED_MODULE_5__src__["a" /* default */], {
            className: 'forTest',
            direction: 'up',
            refreshing: this.state.refreshing,
            onRefresh: function onRefresh() {
              _this4.setState({ refreshing: true });
              setTimeout(function () {
                _this4.setState({ refreshing: false });
              }, 1000);
            }
          })
        },
        [1, 2, 3, 4, 5, 6, 7].map(function (i) {
          return __WEBPACK_IMPORTED_MODULE_6_react___default.a.createElement(
            'div',
            { key: i, style: { textAlign: 'center', padding: 20 } },
            'item ',
            i
          );
        })
      );
    }
  }]);

  return App;
}(__WEBPACK_IMPORTED_MODULE_6_react___default.a.Component);

__WEBPACK_IMPORTED_MODULE_7_react_dom___default.a.render(__WEBPACK_IMPORTED_MODULE_6_react___default.a.createElement(App, null), document.getElementById('__react-content'));

/***/ })

},[109]);
//# sourceMappingURL=base1.js.map

examples/base.js
webpackJsonp([1],{

/***/ 108:
/***/ (function(module, exports, __webpack_require__) {

module.exports = __webpack_require__(54);


/***/ }),

/***/ 54:
/***/ (function(module, __webpack_exports__, __webpack_require__) {

"use strict";
Object.defineProperty(__webpack_exports__, "__esModule", { value: true });
/* harmony import */ var __WEBPACK_IMPORTED_MODULE_0_babel_runtime_helpers_extends__ = __webpack_require__(38);
/* harmony import */ var __WEBPACK_IMPORTED_MODULE_0_babel_runtime_helpers_extends___default = __webpack_require__.n(__WEBPACK_IMPORTED_MODULE_0_babel_runtime_helpers_extends__);
/* harmony import */ var __WEBPACK_IMPORTED_MODULE_1_babel_runtime_helpers_classCallCheck__ = __webpack_require__(20);
/* harmony import */ var __WEBPACK_IMPORTED_MODULE_1_babel_runtime_helpers_classCallCheck___default = __webpack_require__.n(__WEBPACK_IMPORTED_MODULE_1_babel_runtime_helpers_classCallCheck__);
/* harmony import */ var __WEBPACK_IMPORTED_MODULE_2_babel_runtime_helpers_createClass__ = __webpack_require__(21);
/* harmony import */ var __WEBPACK_IMPORTED_MODULE_2_babel_runtime_helpers_createClass___default = __webpack_require__.n(__WEBPACK_IMPORTED_MODULE_2_babel_runtime_helpers_createClass__);
/* harmony import */ var __WEBPACK_IMPORTED_MODULE_3_babel_runtime_helpers_possibleConstructorReturn__ = __webpack_require__(23);
/* harmony import */ var __WEBPACK_IMPORTED_MODULE_3_babel_runtime_helpers_possibleConstructorReturn___default = __webpack_require__.n(__WEBPACK_IMPORTED_MODULE_3_babel_runtime_helpers_possibleConstructorReturn__);
/* harmony import */ var __WEBPACK_IMPORTED_MODULE_4_babel_runtime_helpers_inherits__ = __webpack_require__(22);
/* harmony import */ var __WEBPACK_IMPORTED_MODULE_4_babel_runtime_helpers_inherits___default = __webpack_require__.n(__WEBPACK_IMPORTED_MODULE_4_babel_runtime_helpers_inherits__);
/* harmony import */ var __WEBPACK_IMPORTED_MODULE_5__assets_index_less__ = __webpack_require__(50);
/* harmony import */ var __WEBPACK_IMPORTED_MODULE_5__assets_index_less___default = __webpack_require__.n(__WEBPACK_IMPORTED_MODULE_5__assets_index_less__);
/* harmony import */ var __WEBPACK_IMPORTED_MODULE_6__src__ = __webpack_require__(37);
/* harmony import */ var __WEBPACK_IMPORTED_MODULE_7_react__ = __webpack_require__(13);
/* harmony import */ var __WEBPACK_IMPORTED_MODULE_7_react___default = __webpack_require__.n(__WEBPACK_IMPORTED_MODULE_7_react__);
/* harmony import */ var __WEBPACK_IMPORTED_MODULE_8_react_dom__ = __webpack_require__(52);
/* harmony import */ var __WEBPACK_IMPORTED_MODULE_8_react_dom___default = __webpack_require__.n(__WEBPACK_IMPORTED_MODULE_8_react_dom__);










var App = function (_React$Component) {
  __WEBPACK_IMPORTED_MODULE_4_babel_runtime_helpers_inherits___default()(App, _React$Component);

  function App() {
    var _ref;

    var _temp, _this, _ret;

    __WEBPACK_IMPORTED_MODULE_1_babel_runtime_helpers_classCallCheck___default()(this, App);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = __WEBPACK_IMPORTED_MODULE_3_babel_runtime_helpers_possibleConstructorReturn___default()(this, (_ref = App.__proto__ || Object.getPrototypeOf(App)).call.apply(_ref, [this].concat(args))), _this), _this.state = {
      refreshing: false,
      switchContainer: false
    }, _temp), __WEBPACK_IMPORTED_MODULE_3_babel_runtime_helpers_possibleConstructorReturn___default()(_this, _ret);
  }

  __WEBPACK_IMPORTED_MODULE_2_babel_runtime_helpers_createClass___default()(App, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      // setTimeout(() => { this.setState({ refreshing: true }); }, 10);
      // setTimeout(() => { this.setState({ refreshing: true }); }, 100);
      // setTimeout(() => { this.setState({ refreshing: false }); }, 3000);
    }
  }, {
    key: 'render',
    value: function render() {
      var _this2 = this;

      return __WEBPACK_IMPORTED_MODULE_7_react___default.a.createElement(
        'div',
        null,
        __WEBPACK_IMPORTED_MODULE_7_react___default.a.createElement(
          'button',
          {
            style: { display: 'inline-block', marginBottom: 30, border: 1 },
            onClick: function onClick() {
              return _this2.setState({ switchContainer: !_this2.state.switchContainer });
            }
          },
          'switchContainer: ',
          this.state.switchContainer ? 'true' : 'false'
        ),
        __WEBPACK_IMPORTED_MODULE_7_react___default.a.createElement(
          __WEBPACK_IMPORTED_MODULE_6__src__["a" /* default */],
          __WEBPACK_IMPORTED_MODULE_0_babel_runtime_helpers_extends___default()({
            key: this.state.switchContainer,
            style: { height: 200, overflow: 'auto', border: '1px solid #ccc' }
          }, this.state.switchContainer ? { getScrollContainer: function getScrollContainer() {
              return document.body;
            } } : {}, {
            className: 'forTest',
            direction: 'down',
            refreshing: this.state.refreshing,
            onRefresh: function onRefresh() {
              _this2.setState({ refreshing: true });
              setTimeout(function () {
                _this2.setState({ refreshing: false });
              }, 1000);
            },
            indicator: { deactivate: '下拉' },
            damping: 150
          }),
          [1, 2, 3, 4, 5, 6, 7].map(function (i) {
            return __WEBPACK_IMPORTED_MODULE_7_react___default.a.createElement(
              'div',
              { key: i, style: { textAlign: 'center', padding: 20 }, onClick: function onClick() {
                  return alert(1);
                } },
              'item ',
              i
            );
          })
        ),
        __WEBPACK_IMPORTED_MODULE_7_react___default.a.createElement('div', { dangerouslySetInnerHTML: {
            __html: navigator.userAgent.match(/Android|iPhone|iPad|iPod/i) ? '<style>#qrcode, .highlight{ display: none }</style>' : ''
          }
        })
      );
    }
  }]);

  return App;
}(__WEBPACK_IMPORTED_MODULE_7_react___default.a.Component);

__WEBPACK_IMPORTED_MODULE_8_react_dom___default.a.render(__WEBPACK_IMPORTED_MODULE_7_react___default.a.createElement(App, null), document.getElementById('__react-content'));

/***/ })

},[108]);
//# sourceMappingURL=base.js.map